### PR TITLE
KAFKA-5731 Corrected how the sink task worker updates the last committed offsets

### PR DESCRIPTION
**Pull request by rhauch**
Prior to this change, it was possible for the synchronous consumer commit request to be handled before previously-submitted asynchronous commit requests. If that happened, the out-of-order handlers improperly set the last committed offsets, which then became inconsistent with the offsets the connector task is working with.

This change ensures that the last committed offsets are updated only for the most recent commit request, even if the consumer reorders the calls to the callbacks.

### CODE REVIEW

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6732/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by rhauch** [comment]
The primary change in this PR is to move any logic of setting the `lastOffsetCommits` into the `onCommitCompleted(...)` method, which requires introducing a third parameter for the committed offsets. By doing this, we can reuse the existing logic and update the `lastOffsetCommits` only when the supplied sequence number matches the current sequence number.

However, the new test case to replicate the issue described in KAFKA-5731 also highlighted that after making the aforementioned change did not alone fix the problem. This is because when the consumer signals a rebalance, it did not increment the sequence number. So, the previous invocation of `consumer.commitAsync(...)` would have supplied a commit handler function that used the same sequence number, and that means that even if that handler were called after the rebalance, the handler's invocation of `onCommitCompleted(...)` would have still overwritten the `lastCommittedOffsets`. The solution is for the `WorkerSinkTask` to increment the sequence number in its `onPartitionsAssigned(...)` method, when it also rebuilds the `lastCommittedOffsets`. 

Finally, since an out-of-order `consumer.commitAsync(...)` calls `onCommitCompleted(...)` with an old sequence number, that commit attempt does not reset the worker's `committing` flag. This is normally a good thing, except for our case when a rebalance sneaks in before the async commit's handler, when the task remains in a `committing` state the prevents the next commit from actually occurring. The solution is again for the rebalance to reset the `committing` flag to `false` at the beginning of the `onPartitionsAssigned(...)` method. Essentially, when a rebalance occurs, any in-process commit will not have any effect.

File: checkstyle/suppressions.xml
Comment (rhauch) [inline checkstyle/suppressions.xml:51]: I tried breaking up the new test method in `WorkerSinkTaskTest` into multiple methods, but it actually just made it harder to read.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Note that this does nothing if the supplied `committedOffsets` is null. We use this in most of the invocations of `onCommitCompleted` that are used in the various situations where the commit process does not succeed.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: This is the main invocation of `onCommitCompleted` that is called when the `consumer.commitAsync(...)` method finally calls our `OffsetCommitCallback`, and this change moves the setting of `lastCommittedOffsets` into the `onCommitCompleted` method.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: These two flags are now modified when a rebalance occurs. Incrementing the `commitSeqno` and setting `committing = false` are necessary to effectively ignore any in-flight async commit handler when it calls `onCommitCompleted`. Without this, the in-flight commit handler will call `onCommitCompleted` that _will_ modify and overwrite the correct `lastCommittedOffsets` with the previous generation offsets.

**Comment by rhauch** [comment]
@hachikuji, this is ready for review, and it would be good to port it back at least to 0.11.0.x.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6717/
Test PASSed (JDK 8 and Scala 2.12).

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java
Comment (rhauch) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java:815]: I should note that this assertion fails without the changes to `WorkerSinkTask`, since without those changes the `lastCommittedOffsets` does get overwritten by the `commitAsync`'s handler after the rebalance, and it therefore replicates the problem described in KAFKA-5731.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:368]: No need to fix this here since it would be risky to change, but don't you think it's a little odd that we increment the sequence number even if nothing has changed? I wonder if it would be more intuitive to bump the sequence only prior to the call to `doCommit` below.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: It would be nice to add a short comment about the case when `committedOffsets` is null. I would have expected that only one of `error` and `committedOffsets` would be null, but I see there is a case below where they both are.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Hmm... Is it actually necessary? In the revocation callback, we trigger a synchronous commit which will bump the sequence number. Isn't that sufficient to fence any late commit callbacks?

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java
Comment (hachikuji) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java:649]: This test case is a bit of a pain to read, so I can only imagine how much fun it was to write 😉 . At a glance, I'm wondering if it would have been simpler to capture the commit callbacks so that it's easier to invoke them in the order you want. Thinking longer term, it might be nice to consider refactoring this code to make testing simpler.

**Review by hachikuji — COMMENTED** [review]
Thanks for the patch. Left a few comments/questions.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Ah, maybe my test case doesn't correctly reflect the normal consumer behavior. The consumer will call `onPartitionsRevoked(...)` before it calls `onPartitionsAssigned(...)`?

Comment (hachikuji) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Right. The `onPartitionsRevoked` callback always occurs prior to the rebalance and `onPartitionsAssigned` afterwards.

Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: ok, I'll fix my test case and try removing the four newly-inserted lines.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:368]: Yeah, I wasn't sure why that would be the case, but didn't want to change it if not necessary. I couldn't convince myself there was a situation where it was _wrong_.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerSinkTaskTest.java:649]: Yeah, maybe the test is doing a few things that strictly speaking aren't necessary. For example, the last iteration may not be necessary, but it was only through that following iteration would [this logic](https://github.com/apache/kafka/pull/3662/files/be34cf05fe6956a24b77b35a689bda7b55e7de12#diff-9d27e74bcdc892150367aed9a4cf499eR329) be used after the rebalance. The test is, after all, verifying that a commit, a rebalance, and another commit all work correctly.

**Comment by rhauch** [comment]
@hachikuji, addressed your comments with the most recent 2 commits.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6740/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6755/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by rhauch** [comment]
Retest this please.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Would it make sense to add the offsets to this log message? If so, maybe we could also add them to the other log lines below.

Also, should we consider upgrading the level of this message to INFO and maybe even WARN since this should be a rare case and likely indicates something unexpected has happened?

Comment (hachikuji) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Sorry, one more thing. The message is a little misleading since there wasn't necessarily any timeout. Maybe it should be "{} Received out of order commit callback..."?

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6756/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6741/
Test PASSed (JDK 8 and Scala 2.12).

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: I don't think we should change this message to WARN, since it may happen legitimately until KAFKA-5730 is merged but this PR will address the issue with this situation. Besides, it seems bad form if we were to use WARN but don't include more detail about what users should actually do to correct it.

However, I will add another commit that attempts to improve the log messages as you suggested.

**Comment by rhauch** [comment]
@hachikuji Added another commit with improved log messages.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: nit: could we replace "generation" with "sequence"? The term "generation" is already used by the consumer which makes this message a bit confusing. Same for the other log messages below.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: I wonder if this message and the one in `doCommitSync` is overkill. Maybe we could change the first log message in `doCommit` to include whether it is async or sync. For example?
```java
boolean syncCommit = closing;
log.info("{} Committing {} offsets: {}", this, isSyncCommit? "sync" : "async", offsets);
```

**Review by hachikuji — COMMENTED** [review]
Thanks for the logging improvements. These would have made this problem easier to find. Left a couple suggestions.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6758/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6743/
Test PASSed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6769/
Test FAILed (JDK 8 and Scala 2.12).

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Actually, I pulled out the logic for the asynchronous commit into a `doCommitAsync` method that is similar to the existing `doCommitSync` method, and changed the `doCommit` to call both of them now. Doing this makes the two cases more similar, and I moved the log message into each of the methods.

**Comment by rhauch** [comment]
Retest this please.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6770/
Test FAILed (JDK 8 and Scala 2.12).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6785/
Test PASSed (JDK 7 and Scala 2.11).

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: I think we're missing a `{}`. Same in `doCommitAsync`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Is this log line needed? Seems like we get all this info in `onCommitCompleted`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: I'm not sure we get a useful `toString` from `byte[]`. Maybe it would be more helpful to mention the topic, partition, and offset of the message instead?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Maybe we can combine this log message with the initialization one below? The `subscribe` API does not actually do any IO or anything, so it doesn't seem worth logging separately, though it does seem useful to know what topics are being subscribed to.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: nit (sorry): would be nice to be consistent with capitalization of the first letter.

**Review by hachikuji — COMMENTED** [review]
Thanks. Just a few more comments.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6788/
Test PASSed (JDK 7 and Scala 2.11).

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: What about using `keyAndSchema.value()` and `valueAndSchema.value()` instead of the raw bytes, but also adding the topic, partition, and offset?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: I can change them. Many of the trace messages seem to use lowercase.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Yeah, that might be better.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Up to you I guess. No need to expand the scope even further for a tiny nit.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6774/
Test PASSed (JDK 8 and Scala 2.12).

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (hachikuji) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Did we want to use `valueAndSchema.value()` for the last parameter? Same for the message below.

Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Well, this is a trace-level message. If I were trying to debug a connector, it'd be awesome if I could just turn on trace logging for this class and get a lot more detail about what is going through the worker prior to my task getting it.

But I'll leave it up to you.

Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: I think it would be helpful to have the value, especially if I were trying to watch how the worker and my task handle tombstones or multiple records with the same key.

Comment (hachikuji) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Sure, I'm just saying that `msg.value()` is an array, which does not have a useful `toString` anyway.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6792/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6778/
Test PASSed (JDK 8 and Scala 2.12).

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Yeah. *sigh*

**Review by hachikuji — APPROVED** [review]
LGTM. Thanks for the fix and logging improvements!

**Comment by rhauch** [comment]
@hachikuji, thanks for the reviews! When the builds complete, can you please merge to `trunk` and backport to `0.11.0`? Thanks!!

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk7-scala2.11/6796/
Test PASSed (JDK 7 and Scala 2.11).

**Comment by hachikuji** [comment]
@rhauch Yep, planning to. I'll let the builds finish first for good form.

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-jdk8-scala2.12/6781/
Test PASSed (JDK 8 and Scala 2.12).

**Pull request closed without merging**
